```
Skipping hack/../hack/verify-all.sh
!!! 'gofmt -s' needs to be run on the following files: 
./mungegithub/mungers/submit-queue_test.go
!!! 'golint' problems: 
/home/travis/gopath/src/k8s.io/contrib/mungegithub/mungers/mungerutil/time_cache.go:39:1: exported function NewLabelTimeCache should have comment or be unexported
/home/travis/gopath/src/k8s.io/contrib/mungegithub/mungers/mungerutil/time_cache.go:46:1: exported method LabelTimeCache.FirstLabelTime should have comment or be unexported
```

Fix that and feel free to self LGTM

--- a/mungegithub/github/github.go
+++ b/mungegithub/github/github.go
@@ -274,6 +274,11 @@ type MungeObject struct {
 	Annotations map[string]string //annotations are things you can set yourself.
 }
 
+// Number is short for *obj.Issue.Number.
+func (obj *MungeObject) Number() int {
+	return *obj.Issue.Number
+}
+
 // DebugStats is a structure that tells information about how we have interacted
 // with github
 type DebugStats struct {
@@ -595,8 +600,41 @@ func (obj *MungeObject) LastModifiedTime() *time.Time {
 	return lastModified
 }
 
-// labelEvent returns the most recent event where the given label was added to an issue
-func (obj *MungeObject) labelEvent(label string) *github.IssueEvent {
+// FirstLabelTime returns the first time the request label was added to an issue.
+// If the label was never added you will get a nil time.
+func (obj *MungeObject) FirstLabelTime(label string) *time.Time {
+	event := obj.labelEvent(label, firstTime)
+	if event == nil {
+		return nil
+	}
+	return event.CreatedAt
+}
+
+// Return true if 'a' is preferable to 'b'. Handle nil times!
+type timePred func(a, b *time.Time) bool
+
+func firstTime(a, b *time.Time) bool {
+	if a == nil {
+		return false
+	}
+	if b == nil {
+		return true
+	}
+	return !a.After(*b)
+}
+func lastTime(a, b *time.Time) bool {
+	if a == nil {
+		return false
+	}
+	if b == nil {
+		return true
+	}
+	return a.After(*b)
+}
+
+// labelEvent returns the event where the given label was added to an issue.
+// 'pred' is used to select which label event is chosen if there are multiple.
+func (obj *MungeObject) labelEvent(label string, pred timePred) *github.IssueEvent {
 	var labelTime *time.Time
 	var out *github.IssueEvent
 	events, err := obj.GetEvents()
@@ -606,7 +644,7 @@ func (obj *MungeObject) labelEvent(label string) *github.IssueEvent {
 	index := 0
 	for i, event := range events {
 		if *event.Event == "labeled" && *event.Label.Name == label {
-			if labelTime == nil || event.CreatedAt.After(*labelTime) {
+			if pred(event.CreatedAt, labelTime) {
 				labelTime = event.CreatedAt
 				out = event
 				index = i
@@ -619,9 +657,9 @@ func (obj *MungeObject) labelEvent(label string) *github.IssueEvent {
 }
 
 // LabelTime returns the last time the request label was added to an issue.
-// If the label was never added you will get the 0 time.
+// If the label was never added you will get a nil time.
 func (obj *MungeObject) LabelTime(label string) *time.Time {
-	event := obj.labelEvent(label)
+	event := obj.labelEvent(label, lastTime)
 	if event == nil {
 		return nil
 	}
@@ -630,7 +668,7 @@ func (obj *MungeObject) LabelTime(label string) *time.Time {
 
 // LabelCreator returns the login name of the user who (last) created the given label
 func (obj *MungeObject) LabelCreator(label string) string {
-	event := obj.labelEvent(label)
+	event := obj.labelEvent(label, lastTime)
 	if event == nil || event.Actor == nil || event.Actor.Login == nil {
 		return ""
 	}

--- a/mungegithub/github/testing/github.go
+++ b/mungegithub/github/testing/github.go
@@ -104,6 +104,27 @@ func Issue(user string, number int, labels []string, isPR bool) *github.Issue {
 	return issue
 }
 
+// MultiIssueEvents packages up events for when you have multiple issues in the
+// test server.
+func MultiIssueEvents(issueToEvents map[int][]LabelTime) (out []*github.IssueEvent) {
+	for issueNum, events := range issueToEvents {
+		for _, l := range events {
+			out = append(out, &github.IssueEvent{
+				Issue: &github.Issue{Number: intPtr(issueNum)},
+				Event: stringPtr("labeled"),
+				Label: &github.Label{
+					Name: stringPtr(l.Label),
+				},
+				CreatedAt: timePtr(time.Unix(l.Time, 0)),
+				Actor: &github.User{
+					Login: stringPtr(l.User),
+				},
+			})
+		}
+	}
+	return out
+}
+
 // LabelTime is a struct which can be used to call Events()
 // It expresses what label the event should be about and what time
 // the event took place.
@@ -285,6 +306,19 @@ func setMux(t *testing.T, mux *http.ServeMux, path string, thing interface{}) {
 	})
 }
 
+func splitEventsByIssueNumber(defaultNumber int, events []*github.IssueEvent) map[int][]*github.IssueEvent {
+	// The defaultNumber nonsense is to support tests that were assuming only one issue.
+	out := map[int][]*github.IssueEvent{}
+	for _, e := range events {
+		n := defaultNumber
+		if e.Issue != nil && e.Issue.Number != nil {
+			n = *e.Issue.Number
+		}
+		out[n] = append(out[n], e)
+	}
+	return out
+}
+
 // InitServer will return a github.Client which will talk to httptest.Server,
 // to retrieve information from the http.ServeMux. If an issue, pr, events, or
 // commits are supplied it will repond with those on o/r/
@@ -320,8 +354,10 @@ func InitServer(t *testing.T, issue *github.Issue, pr *github.PullRequest, event
 		setMux(t, mux, path, pr)
 	}
 	if events != nil {
-		path := fmt.Sprintf("/repos/o/r/issues/%d/events", issueNum)
-		setMux(t, mux, path, events)
+		for issueNum, events := range splitEventsByIssueNumber(issueNum, events) {
+			path := fmt.Sprintf("/repos/o/r/issues/%d/events", issueNum)
+			setMux(t, mux, path, events)
+		}
 	}
 	if commits != nil {
 		path := fmt.Sprintf("/repos/o/r/pulls/%d/commits", issueNum)

--- a/mungegithub/mungers/mungerutil/time_cache.go
+++ b/mungegithub/mungers/mungerutil/time_cache.go
@@ -1,0 +1,62 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mungerutil
+
+import (
+	"sync"
+	"time"
+)
+
+// FirstLabelTimeGetter represents anything that can get a label's first time.
+// (interface is for testability / reduced import tree.)
+type FirstLabelTimeGetter interface {
+	FirstLabelTime(label string) *time.Time
+	Number() int
+}
+
+// LabelTimeCache just caches the result of a time lookup, since the first time
+// a label is applied never changes.
+type LabelTimeCache struct {
+	label string
+	cache map[int]time.Time
+	lock  sync.Mutex
+}
+
+// NewLabelTimeCache constructs a label time cache for the given label.
+func NewLabelTimeCache(label string) *LabelTimeCache {
+	return &LabelTimeCache{
+		label: label,
+		cache: map[int]time.Time{},
+	}
+}
+
+// FirstLabelTime returns a time from the cache if possible. Otherwise, it
+// will look up the time. If that doesn't work either, it will return
+// (time.Time{}, false).
+func (c *LabelTimeCache) FirstLabelTime(obj FirstLabelTimeGetter) (time.Time, bool) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if t, ok := c.cache[obj.Number()]; ok {
+		return t, true
+	}
+	t := obj.FirstLabelTime(c.label)
+	if t == nil {
+		return time.Time{}, false
+	}
+	c.cache[obj.Number()] = *t
+	return *t, true
+}

--- a/mungegithub/mungers/mungerutil/time_cache_test.go
+++ b/mungegithub/mungers/mungerutil/time_cache_test.go
@@ -1,0 +1,70 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mungerutil
+
+import (
+	"testing"
+	"time"
+)
+
+type testLTG struct {
+	number int
+	label  string
+	time   time.Time
+}
+
+func (t *testLTG) Number() int { return t.number }
+func (t *testLTG) FirstLabelTime(label string) *time.Time {
+	if label == t.label {
+		return &t.time
+	}
+	return nil
+}
+
+func TestFirstLabelCache(t *testing.T) {
+	timeA := time.Now()
+	timeB := timeA.Add(time.Minute)
+	table := []struct {
+		obj    testLTG
+		expect time.Time
+	}{
+		// Returns the time and caches it.
+		{testLTG{1, "lgtm", timeA}, timeA},
+
+		// Returns zero time, does not cache.
+		{testLTG{2, "blah", timeA}, time.Time{}},
+
+		// Returns old cached value. (Note, in reality the values we
+		// cache should not change.)
+		{testLTG{1, "lgtm", timeB}, timeA},
+
+		// The empty time was not cached.
+		{testLTG{2, "lgtm", timeB}, timeB},
+	}
+
+	cache := NewLabelTimeCache("lgtm")
+	for i, tt := range table {
+		got, ok := cache.FirstLabelTime(&tt.obj)
+		if e := (tt.expect != time.Time{}); e != ok {
+			t.Errorf("%v: Expected %v, got %v", i, e, ok)
+			continue
+		}
+		if got != tt.expect {
+			t.Errorf("%v: Expected %v, got %v", i, tt.expect, got)
+		}
+	}
+}

--- a/mungegithub/mungers/submit-queue.go
+++ b/mungegithub/mungers/submit-queue.go
@@ -184,6 +184,7 @@ type SubmitQueue struct {
 	githubE2ERunning  *github.MungeObject         // protect by sync.Mutex!
 	githubE2EQueue    map[int]*github.MungeObject // protected by sync.Mutex!
 	githubE2EPollTime time.Duration
+	lgtmTimeCache     *mungerutil.LabelTimeCache
 
 	lastE2EStable bool // was e2e stable last time they were checked, protect by sync.Mutex
 	e2e           e2e.E2ETester
@@ -407,6 +408,8 @@ func (sq *SubmitQueue) internalInitialize(config *github.Config, features *featu
 			GoogleGCSBucketUtils: gcs,
 		}).Init(admin.Mux)
 	}
+
+	sq.lgtmTimeCache = mungerutil.NewLabelTimeCache(lgtmLabel)
 
 	if len(config.Address) > 0 {
 		if len(config.WWWRoot) > 0 {
@@ -962,15 +965,18 @@ func priority(obj *github.MungeObject) int {
 	return prio
 }
 
-type queueSorter []*github.MungeObject
+type queueSorter struct {
+	queue          []*github.MungeObject
+	labelTimeCache *mungerutil.LabelTimeCache
+}
 
-func (s queueSorter) Len() int      { return len(s) }
-func (s queueSorter) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s queueSorter) Len() int      { return len(s.queue) }
+func (s queueSorter) Swap(i, j int) { s.queue[i], s.queue[j] = s.queue[j], s.queue[i] }
 
 // If you update the function PLEASE PLEASE PLEASE also update servePriorityInfo()
 func (s queueSorter) Less(i, j int) bool {
-	a := s[i]
-	b := s[j]
+	a := s.queue[i]
+	b := s.queue[j]
 
 	aPrio := priority(a)
 	bPrio := priority(b)
@@ -990,7 +996,20 @@ func (s queueSorter) Less(i, j int) bool {
 		return false
 	}
 
-	return *a.Issue.Number < *b.Issue.Number
+	aTime, aOK := s.labelTimeCache.FirstLabelTime(a)
+	bTime, bOK := s.labelTimeCache.FirstLabelTime(b)
+
+	// Shouldn't really happen since these have been LGTMed to be
+	// in the queue at all. But just in case, .
+	if !aOK && bOK {
+		return false
+	} else if aOK && !bOK {
+		return true
+	} else if !aOK && !bOK {
+		return false
+	}
+
+	return aTime.Before(bTime)
 }
 
 // onQueue just tells if a PR is already on the queue.
@@ -1011,7 +1030,7 @@ func (sq *SubmitQueue) orderedE2EQueue() []int {
 	for _, obj := range sq.githubE2EQueue {
 		prs = append(prs, obj)
 	}
-	sort.Sort(queueSorter(prs))
+	sort.Sort(queueSorter{prs, sq.lgtmTimeCache})
 
 	var ordered []int
 	for _, obj := range prs {
@@ -1305,7 +1324,11 @@ func (sq *SubmitQueue) servePriorityInfo(res http.ResponseWriter, req *http.Requ
       <li>PR with no release milestone will be considered after any PR with a milestone</li>
     </ul>
   </li>
-  <li>PR number</li>
+  <li>First time at which the LGTM label was applied.
+    <ul>
+      <li>This means all PRs start at the bottom of the queue (within their priority and milestone bands, of course) and progress towards the top.</li>
+    </ul>
+  </li>
 </ol> `))
 }
 

--- a/mungegithub/mungers/submit-queue_test.go
+++ b/mungegithub/mungers/submit-queue_test.go
@@ -34,6 +34,7 @@ import (
 	github_test "k8s.io/contrib/mungegithub/github/testing"
 	"k8s.io/contrib/mungegithub/mungers/e2e"
 	fake_e2e "k8s.io/contrib/mungegithub/mungers/e2e/fake"
+	"k8s.io/contrib/mungegithub/mungers/mungerutil"
 	"k8s.io/contrib/test-utils/utils"
 
 	"github.com/golang/glog"
@@ -192,6 +193,7 @@ func getTestSQ(startThreads bool, config *github_util.Config, server *httptest.S
 	sq.lastE2EStable = true
 	sq.prStatus = map[string]submitStatus{}
 	sq.lastPRStatus = map[string]submitStatus{}
+	sq.lgtmTimeCache = mungerutil.NewLabelTimeCache(lgtmLabel)
 
 	sq.startTime = sq.clock.Now()
 	sq.healthHistory = make([]healthRecord, 0)
@@ -211,10 +213,25 @@ func getTestSQ(startThreads bool, config *github_util.Config, server *httptest.S
 }
 
 func TestQueueOrder(t *testing.T) {
+	timeBase := time.Now()
+	time2 := timeBase.Add(6 * time.Minute).Unix()
+	time3 := timeBase.Add(5 * time.Minute).Unix()
+	time4 := timeBase.Add(4 * time.Minute).Unix()
+	time5 := timeBase.Add(3 * time.Minute).Unix()
+	time6 := timeBase.Add(2 * time.Minute).Unix()
+	labelEvents := map[int][]github_test.LabelTime{
+		2: {{"me", lgtmLabel, time2}},
+		3: {{"me", lgtmLabel, time3}},
+		4: {{"me", lgtmLabel, time4}},
+		5: {{"me", lgtmLabel, time5}},
+		6: {{"me", lgtmLabel, time6}},
+	}
+
 	tests := []struct {
-		name     string
-		issues   []*github.Issue
-		expected []int
+		name          string
+		issues        []*github.Issue
+		issueToEvents map[int][]github_test.LabelTime
+		expected      []int
 	}{
 		{
 			name: "Just prNum",
@@ -224,7 +241,8 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, nil, true),
 				github_test.Issue(someUserName, 5, nil, true),
 			},
-			expected: []int{2, 3, 4, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{5, 4, 3, 2},
 		},
 		{
 			name: "With a priority label",
@@ -234,7 +252,8 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, []string{"priority/P0"}, true),
 				github_test.Issue(someUserName, 5, nil, true),
 			},
-			expected: []int{4, 2, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{4, 3, 2, 5},
 		},
 		{
 			name: "With two priority labels",
@@ -244,7 +263,8 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, []string{"priority/P0"}, true),
 				github_test.Issue(someUserName, 5, nil, true),
 			},
-			expected: []int{2, 4, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{4, 2, 3, 5},
 		},
 		{
 			name: "With unrelated labels",
@@ -254,7 +274,8 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, []string{"priority/P0"}, true),
 				github_test.Issue(someUserName, 5, []string{lgtmLabel, "kind/new-api"}, true),
 			},
-			expected: []int{2, 4, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{4, 2, 3, 5},
 		},
 		{
 			name: "With invalid priority label",
@@ -264,7 +285,8 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, []string{"priority/P0", "priorty/bob"}, true),
 				github_test.Issue(someUserName, 5, nil, true),
 			},
-			expected: []int{2, 4, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{4, 2, 3, 5},
 		},
 		{
 			name: "Unlabeled counts as P3",
@@ -274,23 +296,25 @@ func TestQueueOrder(t *testing.T) {
 				github_test.Issue(someUserName, 4, []string{"priority/P2"}, true),
 				github_test.Issue(someUserName, 5, nil, true),
 			},
-			expected: []int{4, 2, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{4, 5, 3, 2},
 		},
 		{
 			name: "retestNotRequiredLabel counts as P-negative 1",
 			issues: []*github.Issue{
 				github_test.Issue(someUserName, 2, nil, true),
 				github_test.Issue(someUserName, 3, []string{"priority/P3"}, true),
-				github_test.Issue(someUserName, 4, []string{"priority/P2"}, true),
+				github_test.Issue(someUserName, 4, []string{"priority/P0"}, true),
 				github_test.Issue(someUserName, 5, nil, true),
 				github_test.Issue(someUserName, 6, []string{"priority/P3", retestNotRequiredLabel}, true),
 			},
-			expected: []int{6, 4, 2, 3, 5},
+			issueToEvents: labelEvents,
+			expected:      []int{6, 4, 5, 3, 2},
 		},
 	}
 	for testNum, test := range tests {
 		config := &github_util.Config{}
-		client, server, mux := github_test.InitServer(t, nil, nil, nil, nil, nil, nil, nil)
+		client, server, mux := github_test.InitServer(t, nil, nil, github_test.MultiIssueEvents(test.issueToEvents), nil, nil, nil, nil)
 		config.Org = "o"
 		config.Project = "r"
 		config.SetClient(client)

